feat(linter): add `eslint/no-misleading-character-class`

Sysix opened the pull request:
closes #614
related #479

https://eslint.org/docs/latest/rules/no-misleading-character-class

The rule is still in nursery because of two problems:
- #13660 
- #13436

The spans for emojis are wrong too, do not know if the rule should fix this, or the regexp parser.

Sysix commented on crates/oxc_linter/src/rules/eslint/no_misleading_character_class.rs:
```diff
@@ -0,0 +1,654 @@
+        (r"/[\u00B7\u0300-\u036F]/u", Some(serde_json::json!([{ "allowEscape": true }]))),
+        (r"/[\n\u0305]/", Some(serde_json::json!([{ "allowEscape": true }]))),
+        // (r#"RegExp("[\uD83D\uDC4D]")"#, Some(serde_json::json!([{ "allowEscape": true }]))),
+        // (r#"RegExp("[A\u0301]")"#, Some(serde_json::json!([{ "allowEscape": true }]))),
```
@leaysgur I expect that this should work because it works with the `RegExpLiteral` in line 448.
The `Character.kind` is `CharacterKind::Symbol` instead of `UnicodeEscape`.

Sysix replied:
https://github.com/oxc-project/oxc/issues/13660#issue-3403241660

camc314 commented on crates/oxc_linter/src/rules/eslint/no_misleading_character_class.rs:
```diff
@@ -0,0 +1,647 @@
+}
+
+// Find regional indicator symbol pairs
+fn regional_indicator_symbol_sequences(chars: &[&Character]) -> bool {
+    for i in 1..chars.len() {
+        let prev = chars[i - 1];
+        let curr = chars[i];
+        if is_regional_indicator_symbol(prev.value) && is_regional_indicator_symbol(curr.value) {
+            return true;
+        }
+    }
+    false
+}
```
nit: could possibly use https://docs.rs/itertools/latest/itertools/structs/struct.TupleWindows.html here

camc314 commented on crates/oxc_linter/src/snapshots/eslint_no_misleading_character_class.snap:
```diff
@@ -0,0 +1,776 @@
+
+  ⚠ eslint(no-misleading-character-class): Unexpected joined character sequence in character class.
+   ╭─[no_misleading_character_class.tsx:1:10]
+ 1 │ var r = /[👨‍👩‍👦]foo[👨‍👩‍👦]/u
+   ·          ───────────────────
```
this span looks funky? it's longer than the ast node.

camc314 replied:
this might cause a panic when diagnostic is generated (probably not since these test snapshots were ok to be generated

camc314 reviewed (approved):
awesome work!

camc314 commented:
> Perhaps there's a problem on the oxc-miette side...?

this seems likley, when rendering the span i guess it doesn't understand that the emojis are amoji sequences

camc314 commented:
Let's fix the spans later

camc314 commented:
### Merge activity

* **Sep 18, 7:59 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Sysix committed:
feat(linter): add `eslint/no-misleading-character-class` (#13365)

closes #614
related #479

https://eslint.org/docs/latest/rules/no-misleading-character-class

The rule is still in nursery because of two problems:
- #13660
- #13436

The spans for emojis are wrong too, do not know if the rule should fix this, or the regexp parser.
